lib/flight_configuration/base_dsl.rb

```diff
@@ -101,22 +105,61 @@ def attribute(name, env_var: true, required: true, default: nil, **opts)
         transform: transform
       }
       attr_accessor name.to_s
+
+      # Define the accessor that returns the original value
+      define_method("_#{name.to_s}") do
+        @__sources__ ||= {}
+        @__sources__[name]&.value
+      end
     end
 
     def load
-      merged = defaults.merge(from_config_files).merge(from_env_vars)
       new.tap do |config|
-        merged.each do |key, value|
+        config.instance_variable_set(:@__sources__, {})
+        merge_sources.each do |key, source|
+          config.instance_variable_get(:@__sources__)[key] = source
           required = attributes.fetch(key, {})[:required]
-          if value.nil? && required
-            raise Error, "The required config has not been provided: #{key}"
-          else
-            config.send("#{key}=", transform(key, value))
+          if source.value.nil? && required
+            if active_errors?
+              config.errors.add(key, :required, message: 'is required')
+            else
+              raise Error, "The required config has not been provided: #{key}"
+            end
+          elsif config.respond_to?("#{key}=")
+            config.send("#{key}=", transform(config, key, source.value))
           end
         end
+
+        # Attempt to validate the config
+        validate!(config)
       end
     rescue => e
-      raise e, "Cannot load configuration:\n#{e.message}", e.backtrace
+      raise e, "Can not continue as the configuration is invalid:\n#{e.message}", e.backtrace
+    end
+
+    def merge_sources
+      {}.tap do |sources|
+        # Apply the env vars
+        from_env_vars.each do |key, value|
+          sources[key] = SourceStruct.new(key, "#{env_var_prefix}_#{key}", :env, value)
+        end
+
+        # Apply the configs
+        config_files.reverse.each do |file|
```

Comment:
It is actually required to maintain the behaviour because the highest priority config is defined last.

Previously `merge` was used, which needs to be ran in low-high priority order. The problem with `merge` is you loose track of which `source` each config came from.

Instead each config needs to be transformed into a `SourceStruct` before compiling/merging into a single hash. This lends itself to preforming the "compile" in reverse-priority order, hence why the `env vars` are interrogated first and defaults last.

It also means the config files need to be applies in the reverse (high-low) priority order. The higher prioity configs are maintained via conditional assignment:

https://github.com/openflighthpc/flight_configuration/blob/68ca0525ef21424b81db1792833a0f24424316f6/lib/flight_configuration/base_dsl.rb#L151

Doing it this way means only `configs` that appear in the final hash will be type casted into a `SourceStruct`.

Reply:
Aha! `next if sources.key?(key)` is the part I'd missed.  Thanks for the explanation.